File: assignment 5/c/programs/7.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for a node in the doubly linked list
struct Node
{
    char data;
    struct Node *prev;
    struct Node *next;
};

// Function to create a new node
struct Node *createNode(char data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    if (newNode == NULL)
    {
        printf("Memory allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the doubly linked list
struct Node *insertAtEnd(struct Node *head, char data)
{
    struct Node *newNode = createNode(data);

    if (head == NULL)
    {
        newNode->next = newNode;
        newNode->prev = newNode;
        return newNode;
    }

    newNode->next = head;
    newNode->prev = head->prev;
    head->prev->next = newNode;
    head->prev = newNode;

    return head;
}

// Function to remove a character from the doubly linked list
struct Node *removeCharacter(struct Node *head, char target)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return NULL;
    }

    struct Node *current = head;

    do
    {
        if (current->data == target)
        {
            // If the node to be removed is the head node
            if (current == head)
            {
                head = head->next;
                head->prev = current->prev;
                current->prev->next = head;
                free(current);
                current = head; // Move to the next node
            }
            else
            {
                struct Node *temp = current;
                current->prev->next = current->next;
                current->next->prev = current->prev;
                current = current->next; // Move to the next node
                free(temp);
            }
        }
        else
        {
            current = current->next;
        }
    } while (current != head);

    return head;
}

// Function to display the doubly linked list
void display(struct Node *head)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return;
    }

    struct Node *current = head;

    do
    {
        printf("%c -> ", current->data);
        current = current->next;
    } while (current != head);

    printf("\n");
}

int main()
{
    struct Node *head = NULL;

    // Initialize a predefined doubly linked list with characters
    head = insertAtEnd(head, 'A');
    head = insertAtEnd(head, 'B');
    head = insertAtEnd(head, 'C');
    head = insertAtEnd(head, 'A');
    head = insertAtEnd(head, 'D');
    head = insertAtEnd(head, 'E');

    printf("Initial Doubly Linked List: ");
    display(head);

    char target;
    printf("Enter the character to remove: ");
    scanf(" %c", &target); // Note the space before %c to consume the newline character

    // Remove the specified character
    head = removeCharacter(head, target);

    printf("Updated Doubly Linked List: ");
    display(head);

    return 0;
}

File: assignment 5/python/programs/7.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the end of the doubly linked list
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = new_node
            new_node.prev = current

    # Remove a particular character from the doubly linked list
    def remove_character(self, char_to_remove):
        if self.head is None:
            print("List is empty. Cannot remove character.")
            return

        current = self.head
        while current:
            if current.data == char_to_remove:
                if current.prev:
                    current.prev.next = current.next
                else:
                    self.head = current.next

                if current.next:
                    current.next.prev = current.prev
                return
            current = current.next

    # Display the doubly linked list
    def display(self):
        if self.head is None:
            print("Doubly Linked List is empty.")
            return

        current = self.head
        while current is not None:
            print(current.data, end=" <-> ")
            current = current.next
        print("None")


if __name__ == "__main__":
    doubly_linked_list = DoublyLinkedList()

    # Insert elements into the doubly linked list
    doubly_linked_list.insert_at_end("A")
    doubly_linked_list.insert_at_end("B")
    doubly_linked_list.insert_at_end("C")
    doubly_linked_list.insert_at_end("D")

    # Display the doubly linked list
    print("Doubly Linked List:")
    doubly_linked_list.display()

    # Remove a particular character from the doubly linked list
    char_to_remove = "B"
    doubly_linked_list.remove_character(char_to_remove)

    # Display the doubly linked list after removal
    print(f"Doubly Linked List after removing '{char_to_remove}':")
    doubly_linked_list.display()
